File: DK/pset1/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

//https://cs50.harvard.edu/x/2020/psets/1/credit/

int main(void)
{
    //request for input credit card number
    long creditNum = get_long("Number: ");

    // count number of digits
    int numLength = 0;
    long n = creditNum;
    while (n != 0)
    {
        n /= 10;
        numLength++;
    }

    //check if input credit card number is an appropriate length
    if (numLength == 13 || numLength == 15 || numLength == 16)
    {
        int numberArray[numLength];
        int c = numLength;
        n = creditNum;

        // extract each digit and place into an array
        while (n != 0)
        {
            numberArray[c - 1] = n % 10;
            n /= 10;
            c--;
        }

        int numTotal = 0;
        //run luhn's algorithm on the input number
        for (int i = (numLength - 2); i >= 0; i -= 2)
        {
            if (numberArray[i] < 5)
            {
                numTotal += numberArray[i] * 2;
            }
            if (numberArray[i] >= 5)
            {
                numTotal += (numberArray[i] - 4) * 2 - 1;
            }
        }

        for (int i = (numLength - 1); i >= 0; i -= 2)
        {
            numTotal += numberArray[i];
        }

        //check if luhn's algorithm is true for the input number
        if (numTotal % 10 == 0)
        {
            //case for AMEX
            if (numLength == 15 && numberArray[0] == 3 && (numberArray[1] == 4 || numberArray[1] == 7))
            {
                printf("AMEX\n");
            }
            //case for MASTERCARD
            else if (numLength == 16 && numberArray[0] == 5 && (numberArray[1] == 1 || numberArray[1] == 2 || numberArray[1] == 3 ||
                     numberArray[1] == 4 || numberArray[1] == 5))
            {
                printf("MASTERCARD\n");
            }
            //case for VISA
            else if ((numLength == 13 || numLength == 16) && numberArray[0] == 4)
            {
                printf("VISA\n");
            }
            //case if none are valid
            else
            {
                printf("INVALID\n");
            }
        }
        //case if luhn's algorithm is not true
        else
        {
            printf("INVALID\n");
        }
    }
    //case if length of input string is not valid (13, 15 or 16)
    else
    {
        printf("INVALID\n");
    }
}

File: DK/pset6/credit/credit.py
# https://cs50.harvard.edu/x/2020/psets/6/credit/


def main():
    # Take user input string
    creditStr = input("Number: ")
    length = len(creditStr)
    
    # Case where length is valid
    if (length == j for j in [13, 15, 16]):
    
        # Case where input is an integer
        try:
            creditNum = int(creditStr)
            # Save reverse of input as a variable
            creditRev = creditStr[::-1]
            totalNum = 0
            # Iterate over every element of input to perform luhn's test
            for i in range(length):
                if i % 2 == 0:
                    totalNum += int(creditRev[i])
                if i % 2 == 1:
                    if int(creditRev[i]) < 5:
                        totalNum += int(creditRev[i]) * 2
                    else:
                        totalNum += ((int(creditRev[i])) - 4) * 2 - 1
            
            # Case where luhn's test suceeds
            if totalNum % 10 == 0:
                if length == 15 and creditStr[0] == '3' and (creditStr[1] == '4' or creditStr[1] == '7'):
                    print("AMEX")
                elif length == 16 and creditStr[0] == '5' and (creditStr[1] == i for i in ['1', '2', '3', '4', '5']):
                    print("MASTERCARD")
                elif (length == 13 or length == 16) and creditStr[0] == '4':
                    print("VISA")
                else:
                    print("INVALID")

            # Case where luhn's test fails
            else:
                print("INVALID")

        # Case where input is not an integer
        except ValueError:
            print("INVALID")

    # Case where length is invalid
    else:
        print("INVALID")


main()
